src/service_data_structure.h

```diff
@@ -0,0 +1,268 @@
+  // returns OK if this operation succeeds
+  // returns other return codes otherwise
+  ReturnCodes ReadChirp(const uint64_t &id, struct Chirp * const chirp);
+};
+
+namespace chirp_connect_backend {
+// Declaration for the `BackendClient` object
+extern std::unique_ptr<BackendClient> backend_client_;
+
+// Binary `User` format:
+// 0x0 - sizeof(struct timeval): struct timeval: last_update_chirp_time
+// Remaining:  username
+//
+// Translate binary data to `struct User`
+void DecomposeBinaryUser(const std::string &input, struct ServiceDataStructure::User * const user);
+// Translate `struct User` to binary data
+void ComposeBinaryUser(const struct ServiceDataStructure::User &user, std::string * const output);
+
+// Binary `following list` format:
+// 0x0 - 0x7 bytes: uint64_t: length of this list
+// Remaining: array of pairs: (username.size(), username)
+//
+// Translate binary data to `UserFollowingList`
+void DecomposeBinaryUserFollowing(const std::string &input, UserFollowingList * const following_list);
+// Translate `UserFollowingList` to binary data
+void ComposeBinaryUserFollowing(const UserFollowingList &following_list, std::string * const output);
+
+// Binary `chirp list` format:
+// 0x0 - 0x7 bytes: uint64_t: length of this list
+// Remaining: uint64_t[]: array of chirp ids
+//
+// Translate binary data to `UserChirpList`
+void DecomposeBinaryUserChirp(const std::string &input, UserChirpList * const chirp_list);
+// Translate `UserChirpList` to binary data
+void ComposeBinaryUserChirp(const UserChirpList &chirp_list, std::string * const output);
+
+// Binary `chirp` format:
+// 0x0 - 0x7 bytes: uint64_t: chirp id
+// 0x8 - 0xF bytes: uint64_t: parent_id
+// 0x10 - 0x10 + sizeof(struct timeval) bytes: struct timeval: time
+//     - +8bytes: uint64_t: length of username
+//     -        : string: username
+//     - +8bytes: uint64_t: length of text
+//     -        : string: text
+//     - +8bytes: uint64_t: length of vector of children chirp ids
+// Remaining: uint64_t[]: array of children chirp ids
+//
+// Translate binary data to `struct Chirp`
+void DecomposeBinaryChirp(const std::string &input, struct ServiceDataStructure::Chirp * const chirp);
+// Translate `struct Chirp` to binary data
+void ComposeBinaryChirp(const struct ServiceDataStructure::Chirp &chirp, std::string * const output);
+
+// Wrapper function to get `next_chirp_id`
+uint64_t GetNextChirpId();
+
+// Wrapper function to get a specified user object
+bool GetUser(const std::string &username, struct ServiceDataStructure::User * const user);
+
+// Wrapper function to save a specified user object
+bool SaveUser(const std::string &username, const struct ServiceDataStructure::User &user);
+
+// Wrapper function to delete a specified user object
+bool DeleteUser(const std::string &username);
+
+// Wrapper function to get the following list of a specified user
+bool GetUserFollowingList(const std::string &username, UserFollowingList * const following_list);
```

Comment:
If you're going to write all of these wrapper functions, I think it might be better if you made User into it's own class; that way you can have accessor fuctions

Reply:
Let me think about it.

Reply:
Finally, I moved all those helper functions with prefix ```Decompose``` and ```Compose``` to their corresponding classes or structs.
However, for those helper functions with prefix ```Get``` and ```Save```, I still put them in a global namespace ```chirp_connect_backend```. There are two reasons. One is that it makes more sense for me to put all functions about connecting to the backend into a separate namespace instead of their own classes. The other is that sometimes for ```Get``` functions, it is only used to check whether the specified object exists, which means it does not involve in creating a object. On the other hand, member functions can be executed only when its object has been created.

The above is my opinion. Still open to discuss.